File: javascript/lesson 14/trappingRainWater.js
// https://leetcode.com/problems/trapping-rain-water/

function trap(height) {
    const leftMax = new Array(height.length);
    leftMax[0] = height[0];
    for (let i = 1; i < height.length; i++) {
        leftMax[i] = Math.max(height[i], leftMax[i - 1]);
    }

    const rightMax = new Array(height.length);
    rightMax[height.length - 1] = height[height.length - 1];
    for (let i = height.length - 2; i >=0; i--) {
        rightMax[i] = Math.max(height[i], rightMax[i + 1]);
    }

    let result = 0;
    for (let i = 1; i < height.length - 1; i++) {
        result += Math.min(leftMax[i], rightMax[i]) - height[i];
    }

    return result;    
}

File: kotlin/lesson 14/TrappingRainWater.kt
// https://leetcode.com/problems/trapping-rain-water/

class Solution {
    fun trap(height: IntArray): Int {
        val leftMax = IntArray(height.size)
        leftMax[0] = height[0]
        for (i in 1..height.size - 1) {
            leftMax[i] = Math.max(height[i], leftMax[i - 1])
        }
        
        val rightMax = IntArray(height.size)
        rightMax[height.size - 1] = height[height.size - 1]
        for (i in height.size - 2 downTo 0) {
            rightMax[i] = Math.max(height[i], rightMax[i + 1])
        }
        
        var result = 0
        for (i in 1..height.size - 1) {
            result += Math.min(leftMax[i], rightMax[i]) - height[i]
        }
        
        return result
    }
}
